source/extensions/clusters/logical_dns/logical_dns_cluster.cc

```diff
@@ -44,23 +46,56 @@ convertPriority(const envoy::config::endpoint::v3::ClusterLoadAssignment& load_a
 }
 } // namespace
 
-LogicalDnsCluster::LogicalDnsCluster(const envoy::config::cluster::v3::Cluster& cluster,
-                                     ClusterFactoryContext& context,
-                                     Network::DnsResolverSharedPtr dns_resolver,
-                                     absl::Status& creation_status)
+absl::StatusOr<std::unique_ptr<LogicalDnsCluster>>
+LogicalDnsCluster::create(const envoy::config::cluster::v3::Cluster& cluster,
```

Comment:
So the following would be equivalent?

```yaml
cluster_type:
  name: "envoy.cluster.dns"
  typed_config:
    "@type": ".../envoy.extensions.clusters.dns"
    logical: true # or some other way to indicate logical vs strict.
    ... # other fields
```

```yaml
cluster_type:
  name: "envoy.cluster.logical_dns"
  typed_config:
    "@type": ".../envoy.extensions.clusters.dns"
    logical: true # this field must be true/is ignored
    ... # other fields
```

But both would use `DnsCluster` under the hood?

Reply:
As I mentioned above, the value in the "@type" field should dictate which plugin to use, not the "name" field, so I think those two configs would be essentially identical.

I think what we want is that the new way to specify a DNS cluster is using the example you gave above (either one), and that maps to a single DNS cluster implementation in Envoy.  For backward compatibility, when we see `type: LOGICAL_DNS`, we convert that to your example with `logical: true`, and if we see `type: STRICT_DNS`, we convert that to your example with `logical: false`.

Reply:
@markdroth Ok, I think we're on the same page now. So to summarize the goals:

* update the DnsCluster proto to include other dns fields (I recall there being a good reason I didn't do this, but can't quite remember. I'll let you know if I reremember it)
* Add a logical/strict switch to DnsCluster
* update the factory code to be based on type urls instead of names
* backcompatibility plumbing.
* **Merge StrictDnsCluster and LogicalDnsCluster into a single DnsCluster class**

Reply:
Yes, that looks right to me!

In the interest of saving you work, it might be a good idea to get positive ack from @wbpcode before you do a lot more work here, just to make sure we're all on board with this approach.

Reply:
Perfect. Looking forward to what @wbpcode has to say.

Reply:
Per above I would personally use a thin wrapper type in the API just in case things diverge. Internally it would be nice to have it use the same type but this is a large enough change that I would probably skip that for now. We could always merge them later.

Reply:
I certainly defer to @mattklein123 in terms of keeping the Envoy implementations separate for now.  But I do think there's an argument to be made that the xDS API surface, which is shared by other data planes, shouldn't need to be constrained by Envoy's current implementation.  gRPC currently supports only LOGICAL_DNS, but if we added STRICT_DNS support, the implementations would be almost identical, to the point where it would make much more sense to have this be an option in a common DNS extension proto rather than have two different extension protos.

I think it should be possible to have a single common DNS extension proto in the xDS API while still keeping the Envoy implementations separate.  This can be done by having a common DNS extension impl that checks the value of the option in the common proto and delegates to either the `LogicalDnsCluster` or `StrictDnsCluster` impls.

@mattklein123, will that work for you?

Reply:
Sure that's fine I'm just worried about there being field creep that don't apply to each other. If we think that would happen an inner option is fine.

Reply:
@markdroth @mattklein123 So the concensus is:

* Hold off on merging the Strict/Logical DNS cluster *implementation*
* Refactor the factories to use type urls instead of name
* Add a logical/strict switch to DnsCluster proto

Reply:
That all looks right!  The only other thing we'd need is the common `DnsCluster` impl, which will basically be a thin shim that will delegate to either `LogicalDnsCluster` or `StrictDnsCluster`.

Also, with respect to "Refactor the factories to use type urls instead of name", I believe the factories are already registered by both proto message type and old-style name, so I think all we really need to do here is change https://github.com/envoyproxy/envoy/blob/c35f0a8f2748b3b582b851933fb42f61688ac729/source/common/upstream/cluster_factory_impl.cc#L46-L56 to do the factory lookup using the proto message type instead of the name.

Reply:
I was thinking of having a shared factory implementation

Reply:
Just to make sure I'm understanding, it sounds like you're suggesting having a single `DnsClusterFactory` that generates either `LogicalDnsCluster` or `StrictDnsCluster` impls, depending on how the field in the config is set?  If so, that seems fine to me, although I'd defer to @mattklein123 on Envoy implementation issues if he has other thoughts.

Reply:
That's fine. I would try to limit the changes to the actual Envoy code and just run things through the legacy shim for now.

Reply:
@markdroth I'm going to suggest that we do refactory by proto message type in another PR. The issue is that none of the clusters are registered by proto message type. Further,  EDS, Original destination, and static clusters do not even have a proto message type. We could

a) give each cluster a message type and register each by its  proto message types ( + backcompatibility plumbing)
b) register clusters that have proto message types by their proto message type and fall back to name until we implement a).

But I'd rather do this in another PR. This one is long enough.

Reply:
I agree that we shouldn't try to convert all of the first-class cluster types to separate extensions in this PR; I think this PR should be focused only on what we need for DNS.

I think we want to do (b) here.  I don't think that should be a lot of work, because if a cluster type is already structured as an extension, then it should be a very small change to register that extension proto by type.  I think we should do that as part of this PR.

Reply:
@markdroth Sorry I just saw this comment. Id much rather put this in another PR and registry clusters by proto because it would allow me to simplify the inheritance chain of all clusters, instead of having the API in a half-baked state.

Reply:
From my other comment

> So confusingly, inheriting from ConfigurableClusterFactoryBase does not do any additional registration. This inheritance only adds some helper methods that parse out the typed_config. The inheritance hierarchy is *DnsClusterFactory -> ConfigurableClusterFactoryBase -> ClusterFactoryImplBase -> ClusterFactory -> UntypedFactory. 

I'm probably overthinking how much of a change this will be. I just want to minimize the diff to make it it easier for whoever is unlucky enough to review this 1000+ line PR, but I'm happy to do it here too.


The reason why we convert from old to new-style configs is because cluster factories are thin wrappers around constructors and `*DnsCluster` constructors take a `v3.Cluster` proto as their main argument. If we really want to avoid making changes to the factories, we would have to have two constructors, one that takes in a `v3.Cluster` proto as the main argument (which exists in main right now) and one also takes in a `DnsCluster` proto. That being said, I'm not too worried about the implementation because I think there's going to be a good amount of plumbing regardless of how we approach it. As long as the API/behavior is clear, I can make it happen.


On the topic of behavior, I just want to make sure we are on the same page

The following will create a logical dns cluster that respects dns ttl (nothing changes)
```yaml
type: LOGICAL_DNS
respect_dns_ttl: true
```

The following will create a logical dns cluster that respects dns ttl (nothing changes)
```yaml
cluster_type:
  name: envoy.cluster.logical_dns
respect_dns_ttl: true
```

The following will create a logical dns cluster that *does not* respect dns ttl
```yaml
cluster_type:
  name: envoy.cluster.logical_dns
  typed_config:
    "@type": ...v3.DnsCluster
respect_dns_ttl: true
```

The following is invalid (note that this works right now, but is probably a good breaking change)
```yaml
cluster_type:
  name: envoy.cluster.logical_dns
  typed_config:
    "@type": ...v3.GetAddrInfoDnsResolverConfig  # something nonsensical
respect_dns_ttl: true
```

The following creates a *strict dns* cluster because `all_addresses_in_single_endpoint` defaults to false and we lookup by config type first (???) . It also doesn't respect dns ttl.
```yaml
cluster_type:
  name: envoy.cluster.logical_dns
  typed_config:
    "@type": ...v3.DnsCluster
respect_dns_ttl: true
```

The following is invalid config because there is no typed config?
```
cluster_type:
  name: envoy.cluster.dns
respect_dns_ttl: true
```

Reply:
> On the topic of behavior, I just want to make sure we are on the same page

Yes, I think the behavior you sketched out here is exactly what we want!

Just to be clear, note that the following should also work, because when we look up by proto message name, the instance name doesn't matter at all:

```
cluster_type:
  name: some_arbitrary_name
  typed_config:
    "@type": ...v3.DnsCluster
```

That would result in a STRICT_DNS cluster.

> The following will create a logical dns cluster that respects dns ttl (nothing changes)
> 
> ```yaml
> cluster_type:
>   name: envoy.cluster.logical_dns
> respect_dns_ttl: true
> ```

Honestly, I didn't even realize prior to this conversation that that case worked today.  I'm fine with retaining backward compatibility in Envoy for this, but I don't think this is a case we want to advertise at the API level, since it's really not the intended behavior, and it certainly isn't something that gRPC supports or will support.

> The following is invalid (note that this works right now, but is probably a good breaking change)
> 
> ```yaml
> cluster_type:
>   name: envoy.cluster.logical_dns
>   typed_config:
>     "@type": ...v3.GetAddrInfoDnsResolverConfig  # something nonsensical
> respect_dns_ttl: true
> ```

That case is even weirder, and my gut feel is to be fine with breaking it, since I doubt anyone would be doing that.  But @mattklein123 should weigh in here -- he may want to be conservative and add a runtime guard to get the old behavior back, just as a migration mechanism.

> I'm probably overthinking how much of a change this will be. I just want to minimize the diff to make it it easier for whoever is unlucky enough to review this 1000+ line PR, but I'm happy to do it here too.

My thinking for preferring it in this PR was that it would be hard to get the behavior described above without fixing the extension registration first.  But if you can find a way to get the behavior described above without doing that, I'm fine with decoupling them.  Another option would be to do the registration change in a separate PR but land that PR first, before this one.

Reply:
Awesome. I'm just curious what your thought process is behind the following

> The following creates a **strict dns cluster** because all_addresses_in_single_endpoint defaults to false and we lookup by config type first **even though the `name` is `envoy.cluster.logical_dns`**.

```yaml
cluster_type:
  name: envoy.cluster.logical_dns
  typed_config:
    "@type": ...v3.DnsCluster
```

On one hand, having the above config create a strict dns cluster is nice because it simplifies the API (check config type, then name), makes it more consistent with other factories, and isn't technically a breaking change. 

On the other hand, I feel like there is a precedence that when `name` is `envoy.cluster.logical_dns` , a logical dns cluster should be created. Another way to say this is the following diff would change a cluster type from logical to strict

```diff
cluster_type:
  name: envoy.cluster.logical_dns
+ typed_config:
+ "@type": ...v3.DnsCluster
```

which could catch someone off guard, even though it wouldn't be technically a breaking change

I'm torn between the two and its probably comes down to what we expect out of users and what users expect out of us, but I'm really curious about your thought process or if there is anything I'm missing

Reply:
I changed this to have it use config proto type if available. If no such config is available, then it errors. If no config proto type is provided, then use name. This will help us reject nonsensical configs like 

```yaml
cluster_type:
  name: envoy.cluster.logical_dns
  typed_config:
    "@type": ...v3.GetAddrInfoDnsResolverConfig  # something nonsensical
```

Reply:
This looks really good to me!